File: vmd_envelope.m
cd('E:\dr.liou\KAIST\1st\vibration');
load("0Nm_BPFO_03.mat");
x = Signal.y_values.values(1:25600, 1);
fs = 25600;
N = length(x);
t = (0:N-1)/fs;

SNR_dB = -20;
x_noisy = awgn(x, SNR_dB, 'measured');

[imfs,residual,info] = vmd(x_noisy, ...
    AbsoluteTolerance=5e-06, ...
    RelativeTolerance=0.005, ...
    MaxIterations=500, ...
    NumIMF=4, ...
    InitialIMFs=zeros(25600,4), ...
    PenaltyFactor=1000, ...
    InitialLM=complex(zeros(25601,1)), ...
    LMUpdateRate=0.01, ...
    InitializeMethod='peaks');

[L,num_imfs] = size(imfs);

for i=1:num_imfs
    IMF = imfs(:,i);
    
    % Envelope
    analytic_signal = hilbert(x); 
    envelope = abs(analytic_signal);

    envelope = envelope - mean(envelope);

    Y = fft(envelope);

    P2 = abs(Y/L);
    P1 = P2(1:floor(L/2)+1);
    P1(2:end-1) = 2*P1(2:end-1);

    f = fs * (0:(L/2)) / L;

    subplot(num_imfs, 1, i);
    plot(f, P1, 'b', 'LineWidth', 1.2);
    title(['Phổ đường bao của IMF ', num2str(i)]);
    xlabel('Tần số (Hz)');
    xlim([0,500]);
    ylabel('Biên độ');
    grid on;
end    

File: vmd_envelope.py
import numpy as np
import scipy.io as sio
from scipy.signal import hilbert
import matplotlib.pyplot as plt
from vmdpy import VMD

# --- Load Data ---
mat_data = sio.loadmat(r'E:\dr.liou\KAIST\1st\vibration\0Nm_BPFI_03.mat')
signal_full = mat_data['Signal'][0, 0]['y_values'][0, 0]['values']
x = signal_full[0:25600, 0]

fs = 25600
N = len(x)
t = np.arange(N) / fs

# --- Add Noise ---
SNR_dB = -5
# Calculate signal power and noise power
sig_power = np.mean(x**2)
snr_linear = 10**(SNR_dB / 10)
noise_power = sig_power / snr_linear
# Generate white Gaussian noise
noise = np.random.normal(0, np.sqrt(noise_power), N)
x_noisy = x + noise

# --- VMD Decomposition ---
alpha = 1000  # Penalty factor   
tau = 0.01    # Time-step of the dual ascent        
K = 4         # Number of modes       
DC = 0        # 0 (no DC part imposed)     
init = 1      # 1 (initializes omegas uniformly)     
tol = 0.005   # Tolerance      

# Run VMD
u, u_hat, omega = VMD(x_noisy, alpha, tau, K, DC, init, tol)

# --- Analysis and Plotting ---
num_imfs = u.shape[0]
L = u.shape[1]

plt.figure(figsize=(10, 12))

for i in range(num_imfs):
    imf = u[i, :]
    
    # Envelope Analysis
    analytic_signal = hilbert(imf) 
    envelope = np.abs(analytic_signal)
    
    # Remove DC component from envelope
    envelope_detrended = envelope - np.mean(envelope)
    
    # Fast Fourier Transform
    Y = np.fft.fft(envelope_detrended)
    
    # Amplitude Spectrum calculation
    P2 = np.abs(Y / L)
    P1 = P2[0 : L // 2 + 1]
    P1[1:-1] = 2 * P1[1:-1]
    
    # Frequency axis
    f = fs * np.arange(L // 2 + 1) / L
    
    # Plotting
    plt.subplot(num_imfs, 1, i + 1)
    plt.plot(f, P1, 'b', linewidth=1.2)
    plt.title(f'Envelope Spectrum of IMF {i+1}')
    plt.xlabel('Frequency (Hz)')
    plt.ylabel('Amplitude')
    plt.xlim([0, 500])
    plt.grid(True)

plt.tight_layout()
plt.show()